src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -245,6 +245,60 @@ bool AbstractLQPNode::is_column_nullable(const ColumnID column_id) const {
   return left_input()->is_column_nullable(column_id);
 }
 
+std::vector<FunctionalDependency> AbstractLQPNode::functional_dependencies() const {
+  // Gather input FDs
+  auto fds_left = std::vector<FunctionalDependency>();
+  auto fds_right = std::vector<FunctionalDependency>();
+  if (left_input()) {
+    fds_left = left_input()->functional_dependencies();
+  }
+  if (right_input()) {
+    fds_right = right_input()->functional_dependencies();
+  }
+
+  auto fds_in = std::vector<FunctionalDependency>();
+  if (fds_right.empty()) {
+    fds_in = fds_left;
+  } else {
+    // Remove duplicate FDs that might result from e.g. self-joins
+    for (const auto& fd_right : fds_right) {
+      bool duplicate = std::any_of(fds_left.begin(), fds_left.end(), [&fd_right](const auto& fd_left) {
+        return (fd_left.first.size() == fd_right.first.size() && fd_left.second.size() == fd_right.second.size() &&
+                fd_left.first == fd_right.first && fd_left.second == fd_right.second);
+      });
+      if (!duplicate) {
+        fds_in.push_back(fd_right);
+      }
+    }
+    std::move(fds_left.begin(), fds_left.end(), std::back_inserter(fds_in));
```

Comment:
We could. This would change line 266 to the following:
```cpp
bool duplicate = std::any_of(fds_in.begin(), fds_in.end(), [&fd_right](const auto& fd_left) { .. }
``` 
Since we add to `fds_in` continuously, the duplicate-check needs additional cycles with for-iteration. But I guess, you consider this as micro-optimization that can be omitted, right?

Reply:
Can't we simply do `fds_in = fds_left;` in all cases and add elements from the right as needed? That would mean that we don't move anymore, but I think the impact of that is negligible.

Reply:
> But I guess, you consider this as micro-optimization that can be omitted, right?

It's good to keep these things in mind. Once readability or debugability are affected, the performance aspect needs to prove its relevance first.